tso-server/src/main/java/org/apache/omid/tso/TSOChannelHandler.java

```diff
@@ -104,21 +100,23 @@ void reconnect() {
         // Create the global ChannelGroup
         channelGroup = new DefaultChannelGroup(TSOChannelHandler.class.getName());
         LOG.debug("\tCreating channel to listening for incoming connections in port {}", config.getPort());
-        listeningChannel = bootstrap.bind(new InetSocketAddress(config.getPort()));
+        Channel listeningChannel = bootstrap.bind(new InetSocketAddress(config.getPort()));
         channelGroup.add(listeningChannel);
         LOG.debug("\tListening channel created and connected: {}", listeningChannel);
     }
 
+    /**
+     * @return the listening channels
+     */
+    Set<Channel> channels() {
+        return channelGroup == null ? Collections.<Channel>emptySet() : new HashSet<>(channelGroup);
```

Comment:
What is the point of copying the channels to a new Hash ?
If you want to block the caller from modfiying it, maybe you could use Collections.unmodifiableSet() ?

Reply:
> maybe you could use Collections.unmodifiableSet() ?

make sense to me.